neqo-transport/src/connection/tests/stream.rs

```diff
@@ -582,6 +618,103 @@ fn simultaneous_stop_sending_and_reset() {
+    client.stream_send(stream_id, &[0x00]).unwrap();
+    let out = client.process_output(now());
+    _ = server.process(out.dgram(), now()).dgram();
+
+    // Make the server generate a packet containing the test frame.
+    let before = server.stats().frame_tx;
+    match frame_type {
+        FRAME_TYPE_RESET_STREAM => {
+            server.stream_reset_send(stream_id, 0).unwrap();
+        }
+        FRAME_TYPE_STOP_SENDING => {
+            server.stream_stop_sending(stream_id, 0).unwrap();
+        }
+        FRAME_TYPE_STREAM_CLIENT_INI_BIDI => {
+            server.stream_send(stream_id, &[0x00]).unwrap();
+            server.stream_close_send(stream_id).unwrap();
+        }
+        FRAME_TYPE_MAX_STREAM_DATA => {
+            server
+                .streams
+                .get_recv_stream_mut(stream_id)
+                .unwrap()
+                .set_stream_max_data(u32::MAX.into());
+        }
+        FRAME_TYPE_STREAM_DATA_BLOCKED => {
+            let internal_stream = server.streams.get_send_stream_mut(stream_id).unwrap();
+            if let SendStreamState::Ready { fc, .. } = internal_stream.state() {
+                fc.blocked();
+            } else {
+                panic!("unexpected stream state");
+            }
+        }
+        _ => panic!("unexpected frame type"),
+    }
+    let tester = server.process_output(now()).dgram();
+    let after = server.stats().frame_tx;
+    match frame_type {
+        FRAME_TYPE_RESET_STREAM => {
+            assert_eq!(after.reset_stream, before.reset_stream + 1);
+        }
+        FRAME_TYPE_STOP_SENDING => {
+            assert_eq!(after.stop_sending, before.stop_sending + 1);
+        }
+        FRAME_TYPE_STREAM_CLIENT_INI_BIDI => {
+            assert_eq!(after.stream, before.stream + 1);
+        }
+        FRAME_TYPE_MAX_STREAM_DATA => {
+            assert_eq!(after.max_stream_data, before.max_stream_data + 1);
+        }
+        FRAME_TYPE_STREAM_DATA_BLOCKED => {
+            assert_eq!(after.stream_data_blocked, before.stream_data_blocked + 1);
+        }
+        _ => panic!("unexpected frame type"),
+    }
+
+    // Now clear the streams on the client, and then deliver the test frame.
+    client.streams.clear_streams();
+    _ = client.process(tester, now()).dgram();
+
+    // Make sure this worked, i.e., the connection didn't close.
+    assert_eq!(*client.state(), State::Confirmed);
+}
+
+#[test]
+fn late_reset_stream_frame() {
+    late_stream_related_frame(FRAME_TYPE_RESET_STREAM);
+}
+
+#[test]
+fn late_stop_sending_frame() {
+    late_stream_related_frame(FRAME_TYPE_STOP_SENDING);
+}
+
+#[test]
+fn late_stream_frame() {
+    late_stream_related_frame(FRAME_TYPE_STREAM_CLIENT_INI_BIDI);
+}
+
+#[test]
+fn late_max_stream_data_frame() {
+    late_stream_related_frame(FRAME_TYPE_MAX_STREAM_DATA);
+}
+
+#[test]
+fn late_stream_data_blocked_frame() {
+    late_stream_related_frame(FRAME_TYPE_STREAM_DATA_BLOCKED);
+}
```

Comment:
I would once again combine these into a single test function. Feel free to ignore.

``` rust
#[test]
fn xxx() {
  for frame in [FRAME_TYPE_RESET_STREAM, ...] {
    late_stream_related_frame(frame)
  }
}

Reply:
Can do. I found separate tests easier to manage while I was debugging `late_stream_related_frame`.